// @aws-cdk/aws-cloudfront/test/oai.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("@aws-cdk/assert/jest");
const cdk = require("@aws-cdk/core");
const lib_1 = require("../lib");
describe('Origin Access Identity', () => {
    test('With automatic comment', () => {
        const stack = new cdk.Stack();
        new lib_1.OriginAccessIdentity(stack, 'OAI');
        expect(stack).toMatchTemplate({
            Resources: {
                OAIE1EFC67F: {
                    Type: 'AWS::CloudFront::CloudFrontOriginAccessIdentity',
                    Properties: {
                        CloudFrontOriginAccessIdentityConfig: {
                            Comment: 'Allows CloudFront to reach the bucket',
                        },
                    },
                },
            },
        });
    });
    test('With provided comment', () => {
        const stack = new cdk.Stack();
        new lib_1.OriginAccessIdentity(stack, 'OAI', {
            comment: 'test comment',
        });
        expect(stack).toMatchTemplate({
            Resources: {
                OAIE1EFC67F: {
                    Type: 'AWS::CloudFront::CloudFrontOriginAccessIdentity',
                    Properties: {
                        CloudFrontOriginAccessIdentityConfig: {
                            Comment: 'test comment',
                        },
                    },
                },
            },
        });
    });
    test('Truncates long comments', () => {
        const stack = new cdk.Stack();
        new lib_1.OriginAccessIdentity(stack, 'OAI', {
            comment: 'This is a really long comment. Auto-generated comments based on ids of origins might sometimes be this long or even longer and that will break',
        });
        expect(stack).toHaveResourceLike('AWS::CloudFront::CloudFrontOriginAccessIdentity', {
            CloudFrontOriginAccessIdentityConfig: {
                Comment: 'This is a really long comment. Auto-generated comments based on ids of origins might sometimes be this long or even longer and t',
            },
        });
    });
    test('Builds ARN of CloudFront user', () => {
        const stack = new cdk.Stack();
        const oai = lib_1.OriginAccessIdentity.fromOriginAccessIdentityName(stack, 'OAI', 'OAITest');
        expect(oai.grantPrincipal.policyFragment.principalJson.AWS[0]).toMatch(/:iam::cloudfront:user\/CloudFront Origin Access Identity OAITest$/);
    });
});
